### DISCUSSION TITLE
Merging reactants by checking reaction SMARTS substructure

### DISCUSSION THREAD

**Discussion by hejops** [rdkit/rdkit#4042] [Q&A]
I have a collection of reactants in the form `A.B` that I need to merge into a single product each, based on a collection of relatively simple reaction/substructure rules (no fancy rearrangements). Below is a MWE of what I have in mind:

```
from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import rdChemReactions

rxn = 'NNN.CCC'
rxn = rdChemReactions.ReactionFromSmarts(f'{rxn}>>', useSmiles=True)

sub = '[N:1].[C:2]>>[N:1][C:2]'		# A
# this SMILES contains the transformation I'm interested in,
# but including the product probably won't work, so...

sub = '[N:1].[C:2]>>'			# B
sub = rdChemReactions.ReactionFromSmarts(sub)

if AllChem.HasReactionSubstructMatch(rxn, sub):
    # do stuff with RunReactants?

...

product = 'CCCNNN'
```

I could probably use `sub.split(">")[0]` to get the reactants part for the substructure check, but is there a more canonical way to accomplish this task?

**Comment by hejops** [comment]
With some messing around, I managed to find a much cleaner solution. I haven't tested it much yet,  but it seems to mostly work for my needs:

```
sub = '[N:1].[C:2]>>[N:1][C:2]'

rxn = 'NNN.CCC'
react1 = rxn.split(".")[0]
react2 = rxn.split(".")[1]
reacts = (Chem.MolFromSmiles(react1), Chem.MolFromSmiles(react2))

products = sub.RunReactants(reacts)

if len(products) == 0:
    continue
```

If anyone knows a better way, let me know!

**Reply by greglandrum** [reply]
The example provided doesn't look like it's going to work since you've reversed the order of the reactants.

Can you provide a bit more detail about what you want to do? The reaction example you should here just forms a single bond between any aliphatic N in reactant 1 and any aliphatic C in reactant 2. I guess that's very oversimplified?

**Reply by hejops** [reply]
I wasn't aware the order of reactants was important; I've edited the example accordingly.

But yes, generally I will be forming one single bond, typically between an aliphatic N and some electrophilic C, ideally with no hydrogens moved.

This is an example I've obtained in my actual use case:

```
sub = '[N:1].[C+:2][C:3]=[C:4]>>[N+:1][C:2][C:3]=[C:4]'
rxn = 'CC(C)N.C1CN2CCCC3=C2C(C1)=CC([CH+]\C=C\C1=CC2=C4N(CCCC4=C1)CCC2)=C3'
pdt
'CC(C)N.C1CN2CCCC3=C2C(C1)=CC([CH+]\C=C\C1=CC2=C4N(CCCC4=C1)CCC2)=C3>>CC(C)[NH2+][C+](/C=C/c1cc2c3c(c1)CCCN3CCC2)c1cc2c3c(c1)CCCN3CCC2'
```

(The `[C+]` in the product should be `C`, but I suspect that's due to lacking sanitisation or some error in my smarts; I'll have to figure that out)

I have a list of `rxn`s to iterate through, as well as a list of `sub`s. For each `rxn`, each `sub` will be tried, and any possible products will be returned.

Hope this makes sense.

**Reply by greglandrum** [reply]
So you have a transformation describing the reaction: `[N:1].[C+:2][C:3]=[C:4]>>[N+:1][C:2][C:3]=[C:4]`
and you would like to end up with a reaction SMILES which includes the actual reactants and product(s) from applying that transformation to the given reactants?

Note: to get the RDKit to change the postively charged C in the reactants to a neutral C in the products, you need to explicitly tell it to remove the charge: `[N:1].[C+:2][C:3]=[C:4]>>[N+:1][C+0:2][C:3]=[C:4]`

**Reply by hejops** [reply]
Yep, exactly. There are a few more transformations that I'll be looking for, but I haven't devised the smarts for them quite yet... Also, thanks for the note about the charge!
